File: Phoenix/lib/bd_starter_web/controllers/subscription_controller.ex
defmodule BdStarterWeb.SubscriptionController do
  use BdStarterWeb, :controller

  alias BdStarter.Bd
  alias BdStarter.Bd.Resources.Subscriptions

  def index(conn, _params), do: render(conn, :index, plans: Bd.subscription_plans())

  def checkout(conn, %{"id" => id}) do
    urls = %{
      successUrl: url(~p"/subscriptions") <> "?status=success",
      cancelUrl: url(~p"/subscriptions") <> "?status=cancelled"
    }

    result =
      case Bd.client() do
        nil -> nil
        client -> Subscriptions.checkout(client, id, urls)
      end

    case result do
      {:ok, %{"stripeUrl" => stripe_url}} ->
        redirect(conn, external: stripe_url)

      _ ->
        conn
        |> put_flash(:error, "Could not start checkout.")
        |> redirect(to: ~p"/subscriptions")
    end
  end
end

File: Laravel/app/Http/Controllers/SubscriptionController.php
<?php

namespace App\Http\Controllers;

use App\Bd\Bd;
use App\Bd\Client;
use Illuminate\Http\RedirectResponse;
use Illuminate\View\View;

class SubscriptionController extends Controller
{
    public function index(): View
    {
        $plans = Bd::remember(
            'subscriptions:list',
            ['bd:subscriptions'],
            static fn (Client $c) => $c->subscriptions()->list(),
            default: ['items' => []],
        );

        return view('pages.subscriptions', [
            'plans' => $plans['items'] ?? [],
        ]);
    }

    public function checkout(string $id): RedirectResponse
    {
        $session = Bd::attempt(static fn (Client $c) => $c->subscriptions()->checkout($id, [
            'successUrl' => route('subscriptions').'?status=success',
            'cancelUrl' => route('subscriptions').'?status=cancelled',
        ]));

        $url = $session['stripeUrl'] ?? $session['url'] ?? null;
        if (! is_string($url)) {
            return back()->with('subscription_error', 'Could not start checkout.');
        }

        return redirect()->away($url, 303);
    }
}
